src/Chainweb/Cut/Create.hs

```diff
@@ -0,0 +1,442 @@
+-- adjacent parent hashes are consistent with the cut.
+--
+-- Only those parents are included that are not block parent hashes of genesis
+-- blocks. This can occur when new graphs are introduced. The headers are used
+-- in 'newBlockHeader' for computing the target. That means that during a graph
+-- change the target computation may only use some (or none) adjacent parents to
+-- adjust the epoch start, which is fine.
+--
+-- If the parent header doesn't exist, because the chain is new, only the
+-- genesis parent hash is included as 'Left' value.
+--
+getAdjacentParentHeaders
+    :: HasCallStack
+    => Applicative m
+    => (ChainValue BlockHash -> m BlockHeader)
+    -> CutExtension
+    -> m (HM.HashMap ChainId (Either BlockHash ParentHeader))
+getAdjacentParentHeaders hdb extension
+    = itraverse select
+    . _getBlockHashRecord
+    $ _cutExtensionAdjacentHashes extension
+  where
+    ver = _chainwebVersion c
+    c = _cutExtensionCut extension
+
+    select cid h = case c ^? ixg cid of
+        Just ch -> Right . ParentHeader <$> if _blockHash ch == h
+            then pure ch
+            else hdb (ChainValue cid h)
+
+        Nothing | h == genesisParentBlockHash ver cid -> Left <$> pure h
+
+        Nothing -> error $ T.unpack
+            $ "Chainweb.Cut.Create.getAdjacentParentHeaders: inconsistent cut extension detected"
+            <> ". ChainId: " <> encodeToText cid
+            <> ". CutHashes: " <> encodeToText (cutToCutHashes Nothing c)
+
+-- -------------------------------------------------------------------------- --
+-- Solved Header
+--
+
+-- | A solved header is the result of mining. It has the final creation time and
+-- a valid nonce. The binary representation doesn't yet contain a Merkle Hash.
+-- Also, the block header is not yet validated.
+--
+-- This is an internal data type. On the client/miner side only the serialized
+-- representation is used.
+--
+newtype SolvedWork = SolvedWork BlockHeader
+    deriving (Show, Eq, Ord, Generic)
+    deriving anyclass (NFData)
+
+encodeSolvedWork :: MonadPut m => SolvedWork -> m ()
+encodeSolvedWork (SolvedWork hdr) = encodeBlockHeaderWithoutHash hdr
+
+decodeSolvedWork :: MonadGet m => m SolvedWork
+decodeSolvedWork = SolvedWork <$> decodeBlockHeaderWithoutHash
+
+data InvalidSolvedHeader = InvalidSolvedHeader BlockHeader T.Text
+    deriving (Show, Eq, Ord, Generic)
+    deriving anyclass (NFData)
+
+instance Exception InvalidSolvedHeader
+
+-- | Extend a Cut with a solved work value.
+--
+-- The function verifies that the solved work actually is a valid extension of
+-- the cut and matches the given payload data. It also checks a few other
+-- invariants. It does't do a full validation.
+--
+-- The resulting 'CutHashes' value contains the new header and payload as
+-- attachement and can be submitted into a cut pipeline.
+--
+-- The result is 'Nothing' if the given cut can't be extended with the solved
+-- work.
+--
+extend
```

Comment:
```suggestion
extendCut
```

Reply:
(and obviously rename internal function to `extendCutInternal` or whatever)

Reply:
I like to defer this to the above mentioned follow up PR.